### PR TITLE
Finish All README writing and some bug fixes

### LINKED ISSUE #115: Add README for backend API
Tasks to do:
- [ ] Create a README for auth-service, include information about how to set up, and the expected .env variables, if possible, add API docs as well, tbd by @tlyi / @xingjie99 
- [x] Create a README for collaboration-service, include information about how to set up, and the expected .env variables, if possible, add events docs as well, tbd by @weidak (added in #103) 
- [x] Create a README for matching-service, include information about how to set up, and the expected .env variables, if possible, add API docs as well, tbd by @c0j0s (added in #113)
- [ ] Update README for question-service, include information about how to set up, and the expected .env variables, as well as the API docs, tbd by @tryyang2001 
- [ ] Update README for user-service, include information about how to set up, and the expected .env variables, as well as the API docs, tbd by @tryyang2001 
- [x] Create a README for history-service, include information about how to set up, and the expected .env variables, as well as the API docs, tbd by @tryyang2001 (added in #104 )

### LINKED ISSUE #133: Update User Service `getUserById` endpoint to not return secrets
## The problem

Currently, auth service is using user service `getUserById` (and perhaps more endpoints) to verify if a user is valid, password matching, and etc. However, this is not a good practice. The issue is that we actually call `getUserById` a lot of time and we are not using password, passwordResetToken, etc. 
```json
{
  "id": "Your Id",
  "name": "Your Name",
  "email": "youremail@domain.com",
  "role": "USER",
  "image": "http://your-image.png",
  "bio": "Your Bio",
  "gender": "MALE",
  "createdOn": "2023-10-17T15:30:30.751Z",
  "updatedOn": "2023-10-27T09:31:02.105Z",
  "isVerified": true,
  "password": "Your Password..." 
  "verificationToken": "",
  "passwordResetToken": null,
  "preferences": {
    "languages": [
      "C++"
    ],
    "topics": [
      "DYNAMIC PROGRAMMING"
    ],
    "difficulties": [
      "MEDIUM",
      "EASY"
    ]
  }
}
```

So, I plan to update `getUserById` to not return `password`, `verificationToken`, and `passwordResetToken`. I will update auth service and frontend accordingly as needed. The updated `getUserById` endpoint response should only return:
```json
{
  "id": "clmlp93wz00007kbwvws8oynd",
  "name": "Your Nae",
  "email": "youremail@domain.com",
  "role": "USER",
  "image": "http://your-image.png",
  "bio": "Your Bio",
  "gender": "MALE",
  "createdOn": "2023-10-17T15:30:30.751Z",
  "updatedOn": "2023-10-27T09:31:02.105Z",
  "isVerified": true,
  "preferences": {
    "languages": [
      "C++"
    ],
    "topics": [
      "DYNAMIC PROGRAMMING"
    ],
    "difficulties": [
      "MEDIUM",
      "EASY"
    ]
  }
}
```

### PR DESCRIPTION
**Pull request by tryyang2001**
# Pull Request

## Description

Update README for user service and history service, create README for question service. Also fixed bugs found when doing the development, improve security and enhance auth code logic. Below are some problems that have been solved:
- [x] Update `postToHistoryService` code logic. In the previous PR, I only try to POST /history but never try the PUT /history. So the issue is that if the user has already attempted the question, and trying to re-post the history will cause a 409 and then failed to store the latest code. So if we get 409, we should call PUT /history/user/:userId/question/:questionId/code instead to save the code submission to the database.
- [x] Update `EndSessionModal` such that if the user tried to terminate session without modifying the code (this will imply the code is `null` or `empty`), the code submission will not be saved and the question will not be marked as attempted.
- [x] `getUserById` and `getUserByEmail` should not return password and secret tokens. This is mainly for security, updated! The chain reaction is that the auth endpoints returned data (user information) and the frontend API call logic also need an update. I have tested login, forgot password, and change password, they are working as expected even after the change.
- [x] Missing error handling in auth APIs. For example, we never check if a request body is valid, properties that should not be NULL/undefined, and many other potential issues. I have gone through all the auth endpoints one by one, and added some error handling to prevent the server crash. Will keep on improving the API when I start creating unit tests for auth.
(@tlyi ) To be done:
- [x] Some auth logic flaws. One of these issues is that we are currently generating the email verification token in the user service, and use auth service to send the email only. Why not we let auth service handles the token generation and email sending? This is the same for forgot password token generation. I have discussed with @tlyi and she will be working on the change!  
- [x] Password reset bug where "used" email is able to reset password
- [x] Auth README

## Related Issue(s)

Closes #115 
Closes #133

## Changes Made

- refer to description
- update README
- update auth service code
- update frontend integration with auth
- update error page to include a `Back to Dashboard` button
- update `ResetPassword` page input field placeholder to make it more descriptive
- update postToHistory code logic
- code refactor, improve efficiency (for eg, check those sync operation first before checking any async operation)
- shift issuing of email verification token from user service to auth service, this means that user service env file no longer need the email secret 
- fix password reset bug where old emails are able to change the password. now, frontend will immediately show error too if the link redirected from the email is invalid (backend checks for this as well)
- update all the languages to render capitalised lower case instead of all uppercase
- add tooltips for question history link on hover
- allow user to enter Edit Profile page by clicking their profile avatar in the dashboard profile card

## Screenshots (if applicable)

An example of the error page will now be:
image

If link from password reset email is invalid:
image

The dashboard UI after the change:
image

Question history link on hover:
image

Profile Avatar on hover (on click will go to the profile page):
image

## Checklist
- [x] I have checked that the changes included in the PR are intended to merge to `master` or any destination branch.
- [x] I have verified that the new changes do not break any existing functionalities, unless the new changes are intended and have approved by the team.
- [x] I will take care of the merging and delete the side-branch after the PR is merged.

### CODE REVIEW

**Comment by weidak** [comment]
Added commit to fix 4. mentioned in #137

**Comment by weidak** [comment]
Not sure if u should fix it here but i found a bug in history service:

image

It renders today (Oct 30) andtomorrow (Oct 31) so when i complete the qn its gonna append to that white box. Think we can remove tomorrow's box?

**Comment by tryyang2001** [comment]
Oh how can you append submission to the future date? I was thinking about the use case. 
But in terms of removing it, unfortunately I can't. I have no control to decide how many days I should render per month. I can only decide how many months I should render... If I have time I will try re-draw the heatmap to render by days and format according to the month, but it takes quite some time...

**Comment by tryyang2001** [comment]
@tlyi looks like your new commits break user service unit tests, can you help resolve the errors?

**Comment by tlyi** [comment]
> @tlyi looks like your new commits break user service unit tests, can you help resolve the errors?

Oops!! Updated, thanks :D

**Comment by tryyang2001** [comment]
Thanks @tlyi for the collaboration in writing README! They look super neat and clear now!

**Comment by tryyang2001** [comment]
The work for this PR has been done, please help review it when you have time. Thanks!

File: backend/user-service/src/routes/index.ts [hunk 1]
```diff
@@ -28,10 +26,6 @@ router.route("/users").post(postUser);
 
 router.route("/users/:userId").put(updateUserById);
 
-router.route("/users/updateVerification/:email").put(updateVerification)
-
-router.route("/users/updatePasswordResetToken/:email").put(updatePasswordResetToken)
-
```
Comment (c0j0s) [inline backend/user-service/src/routes/index.ts:34]: Same as auth-service.

File: backend/auth-service/src/routes/index.ts [hunk 2]
```diff
@@ -12,50 +15,64 @@ import {
...
+
 router.route("/logout").post(logOut);
+
 router.route("/verifyEmail/:email/:token").put(verifyUserEmail);
+
 router.route("/sendPasswordResetEmail/:email").put(sendPasswordResetEmail);
+
+router
+  .route("/verifyResetPasswordLinkValidity/:id/:token")
+  .get(verifyResetPasswordLinkValidity);
```
Comment (c0j0s) [inline backend/auth-service/src/routes/index.ts:74]: Might need to reflect the change in AWS stack template.

File: .github/workflows/deploy_backend_aws_dev.yml [hunk 3]
```diff
@@ -84,6 +84,7 @@ jobs:
             EMAIL_VERIFICATION_SECRET=${{ secrets.EMAIL_VERIFICATION_SECRET }}
             NM_MAIL=${{ secrets.NM_MAIL }}
             NM_PASS=${{ secrets.NM_PASS }}
+            CLOUD_URL=${{ vars.FRONTEND_URL }}
```
Comment (c0j0s) [inline .github/workflows/deploy_backend_aws_dev.yml:88]: Seems like you'll need `DATABASE_URL` here now.

**Review by c0j0s — COMMENTED** [review]
issue:
1. ~~404 with history service when terminating collab, service is live. am I missing any new configs?~~

suggestion:
- [ ] possible to display user entered mail and add a resent verification mail button here? seems like I didn't receive the mail, will be gd for user to see the email addr they have entered and allow resent to destress the situation. 
image

minor
1. ~~ocd~~
3. ~~since we are improving our UI, I think this can be better? 🙃~~

**Comment by tlyi** [comment]
> 1. possible to display user entered mail and add a resent verification mail button here? seems like I didn't receive the mail, will be gd for user to see the email addr they have entered and allow resent to destress the situation.

Good idea! I have implemented this feature! As for validity of verification email, now only the latest email will work :) 
This is what the user will see now when they sign up: 
image

File: .github/workflows/deploy_backend_aws_dev.yml (hunk 3 above)
Comment (tlyi) [reply .github/workflows/deploy_backend_aws_dev.yml]: Ahh, thanks for catching all these! Have updated all the changes locally. Will update the AWS stack template soon! 👍

**Comment by tryyang2001** [comment]
> 404 with history service when terminating collab, service is live. am I missing any new configs?

Can you elaborate more on how to reproduce this error? You will get 404 if the user id and question id provided cannot be found in the database. It is independent of the collab service.

**Comment by c0j0s** [comment]
> Also, just checking, did you actually not receive the email? :O was there no error toast?

1. I can receive the mail now for another account, but there's another issue, if the user did not verify the email due to whatever reasons, the email address is forever locked in the system, and they cannot sign up again to get a new verification mail using the same address.
2. not sure if it's related, but possible to set the default user preference in their profile when signup? just to make the out-of-the-box experience effortless.

**Comment by c0j0s** [comment]
> Can you elaborate more on how to reproduce this error? You will get 404 if the user id and question id provided cannot be found in the database. It is independent of the collab service.
> 

Happens when I tries to terminate the collab.
image
I have the envs.
image

user and question id should be valid since I'm login and already in a collab session.

**Comment by tryyang2001** [comment]
The db should be same as our production database url (the one without -dev). I think your one should be correct. I will try to reproduce this error on my side and see if I can fix this. From your description, I highly doubt that it is because I am not receiving the right user id and question id in the request body that is constructed in the frontend. It may or may not be related to my changes hmm...

**Comment by tryyang2001** [comment]
@c0j0s I can't reproduce the 404 error on my side though, it works perfectly for me.

**Comment by c0j0s** [comment]
> 
> Also, can I confirm if you have this env var in your frontend .env:
> 

image

while I'm at it, don't mind If I push a small UI patch here 😊

**Comment by tryyang2001** [comment]
> don't mind If I push a small UI patch here

Sure, go ahead.

File: .github/workflows/deploy_backend_aws_dev.yml [hunk 4]
```diff
@@ -172,42 +144,42 @@ jobs:
...
+              ecs_service: peerprep-development-ServiceCollaboration,
+              ecs_task_def: peerprep-development-TaskDefinitionCollaboration,
+            },
+            {
+              name: event-bus,
+              ecs_service: peerprep-development-ServiceEventBus,
+              ecs_task_def: peerprep-development-TaskDefinitionEventBus,
+            },
+            # {
+            #   name: history-service,
```
Comment (weidak) [inline .github/workflows/deploy_backend_aws_dev.yml:178]: :O should we uncomment this?

**Review by weidak — APPROVED** [review]
Thank you guys so much for the README! Hopefully the mentors will be able to find them... 😄 This looks good, love the new UI tweaks.. 

> Question history link on hover:

Hopefully the button works on amplify after merge? If not we'll have to tweak it slightly.

**Comment by tryyang2001** [comment]
> Hopefully the button works on amplify after merge?

Just curious, why will it not work?

**Comment by weidak** [comment]
> > Hopefully the button works on amplify after merge?
> 
> Just curious, why will it not work?

Mentioned by @c0j0s in #142

**Comment by tlyi** [comment]
> they cannot sign up again to get a new verification mail using the same address.

Oh okay I will allow unverified uses to request for another email when they try to log in! Maybe in my next PR with code exec ? If @tryyang2001 wants to close this PR soon :) 

> not sure if it's related, but possible to set the default user preference in their profile when signup? just to make the out-of-the-box experience effortless.

Yes it's possible, but what should we set it to? O_O

**Comment by tryyang2001** [comment]
Anyone has any progress left? If not I will merge this branch at 5PM.

**Comment by c0j0s** [comment]
> > they cannot sign up again to get a new verification mail using the same address.
> 
> Oh okay I will allow unverified uses to request for another email when they try to log in! Maybe in my next PR with code exec ? If @tryyang2001 wants to close this PR soon :)
> 
> > not sure if it's related, but possible to set the default user preference in their profile when signup? just to make the out-of-the-box experience effortless.
> 
> Oh you mean in the sign up form itself?

1. sure, let's merge this pr first.
2. the easiest is we prefill it for them backend with the first value from the list🙃

**Comment by weidak** [comment]
> :O should we uncomment this?

@tlyi @tryyang2001 why was history service's deployment code commented out again? Is there a reason why we are not deploying?

**Comment by tryyang2001** [comment]
Hmm, the issue is caused by me in this [commit](https://github.com/CS3219-AY2324S1/ay2324s1-course-assessment-g05/pull/135/commits/93788f431ad8bb51ad803abfd2039846210b61b0). Something weird is that when I pulled from master, there is no merge conflict for this branch, that's why I don't even notice the yaml files versioning is wrong. Sorry, I will create a new branch to roll it back.

**Pull request merged**